src/map/clif.cpp

```diff
@@ -6126,97 +6126,79 @@ void clif_cooking_list( struct map_session_data *sd, int trigger, uint16 skill_i
 /// 0983 <index>.W <id>.L <state>.B <total msec>.L <remain msec>.L { <val>.L }*3 (ZC_MSG_STATE_CHANGE3) (PACKETVER >= 20120618)
-/// @param bl Sends packet to clients around this object
-/// @param id ID of object that has this effect
-/// @param type Status icon see enum efst_types
-/// @param flag 1:Active, 0:Deactive
-/// @param tick Duration in ms
-/// @param val1
-/// @param val2
-/// @param val3
-void clif_status_change_sub(struct block_list *bl, int id, int type, int flag, t_tick tick, int val1, int val2, int val3, enum send_target target_type)
+/// @param bl: Sends packet to clients around this object
+/// @param type: Status icon (see efst_types)
+/// @param flag: 1:Active, 0:Inactive
+/// @param tick_total: Total duration in ms
+/// @param tick: Remaining duration in ms
+/// @param val1: Value 1
+/// @param val2: Value 2
+/// @param val3: Value 3
+void clif_status_change_sub(struct block_list *bl, int type, int flag, t_tick tick_total, t_tick tick, int val1, int val2, int val3)
 {
-	unsigned char buf[32];
+	nullpo_retv(bl);
 
 	if (type == EFST_BLANK)  //It shows nothing on the client...
 		return;
 
 	if (type == EFST_POSTDELAY && tick == 0)
 		return;
 
-	nullpo_retv(bl);
+	if (!(status_type2relevant_bl_types(type) & bl->type)) // only send status changes that actually matter to the client
+		return;
 
-	// Statuses with an infinite duration, but still needs a duration sent to display properly.
-	if (type == EFST_LUNARSTANCE || type == EFST_UNIVERSESTANCE || type == EFST_SUNSTANCE || type == EFST_STARSTANCE)
-		tick = 200;
+	packet_status_change p = { 0 };
+	map_session_data *sd = BL_CAST(BL_PC, bl);
 
-#if PACKETVER >= 20120618
-	if (flag && battle_config.display_status_timers)
-		WBUFW(buf,0) = 0x983;
-	else
-#elif PACKETVER >= 20090121
-	if (flag && battle_config.display_status_timers)
-		WBUFW(buf,0) = 0x43f;
-	else
-#endif
-		WBUFW(buf,0) = 0x196;
-	WBUFW(buf,2) = type;
-	WBUFL(buf,4) = id;
-	WBUFB(buf,8) = flag;
-#if PACKETVER >= 20120618
-	if (flag && battle_config.display_status_timers) {
-		if (tick <= 0)
-			tick = 9999; // this is indeed what official servers do
+	if (battle_config.display_status_timers > 0) {
+		// Statuses with an infinite duration, but still needs a duration sent to display properly.
+		switch (type) {
+			case EFST_LUNARSTANCE:
+			case EFST_UNIVERSESTANCE:
+			case EFST_SUNSTANCE:
+			case EFST_STARSTANCE:
+				tick = 200;
+				break;
+			default:
+				if (tick < 0)
+					tick = 9999; // this is indeed what official servers do
+				break;
+		}
 
-		WBUFL(buf,9) = client_tick(tick);/* at this stage remain and total are the same value I believe */
-		WBUFL(buf,13) = client_tick(tick);
-		WBUFL(buf,17) = val1;
-		WBUFL(buf,21) = val2;
-		WBUFL(buf,25) = val3;
-	}
-#elif PACKETVER >= 20090121
-	if (flag && battle_config.display_status_timers) {
-		if (tick <= 0)
-			tick = 9999; // this is indeed what official servers do
+		p.PacketType = status_changeType;
+	} else
+		p.PacketType = sc_notickType;
```

Comment:
This uses a dangerous assumption and should be changed.
It reintreprets the "newer" packet structure as the old one.
Additionally since we use the sizeof operator on the structure it will be the old packet id with the new packet size and could cause problems.